src/callback_handler/kubernetes.rs

```diff
@@ -0,0 +1,201 @@
+use anyhow::{anyhow, Result};
+use cached::proc_macro::cached;
+use kube::api::ListParams;
+use serde::Serialize;
+use std::collections::{hash_map, HashMap};
+
+#[derive(Eq, Hash, PartialEq)]
+struct ApiVersionKind {
+    api_version: String,
+    kind: String,
+}
+
+#[derive(Debug, Clone, Serialize)]
+struct KubeResource {
+    pub resource: kube::api::ApiResource,
+    pub namespaced: bool,
+}
+
+pub(crate) struct Client {
+    kube_client: kube::Client,
+    kube_resources: HashMap<ApiVersionKind, KubeResource>,
+}
+
+/// This is a specialized `kube::api::ObjectList` object which
+/// implements the `Clone` trait. This trait is required by
+/// `cached::Return<>`
+#[derive(Clone, Serialize)]
+pub(crate) struct ObjectList {
+    pub metadata: kube::core::ListMeta,
+    pub items: Vec<kube::core::DynamicObject>,
+}
+
+impl Client {
+    pub fn new(client: kube::Client) -> Self {
+        Self {
+            kube_client: client,
+            kube_resources: HashMap::new(),
+        }
+    }
+
+    /// Build a KubeResource using the apiVersion and Kind "coordinates" provided.
+    /// The result is then cached locally to avoid further interactions with
+    /// the Kubernetes API Server
+    async fn build_kube_resource(&mut self, api_version: &str, kind: &str) -> Result<KubeResource> {
+        let avk = ApiVersionKind {
+            api_version: api_version.to_string(),
+            kind: kind.to_string(),
+        };
+
+        let kube_resource = match self.kube_resources.entry(avk) {
+            hash_map::Entry::Vacant(entry) => {
+                let resources_list = match api_version {
+                    "v1" => {
+                        self.kube_client
+                            .list_core_api_resources(api_version)
+                            .await?
+                    }
+                    _ => {
+                        self.kube_client
+                            .list_api_group_resources(api_version)
+                            .await?
+                    }
+                };
+
+                let resource = resources_list
+                    .resources
+                    .iter()
+                    .find(|r| r.kind == kind)
+                    .ok_or_else(|| anyhow!("Cannot find resource {api_version}/{kind}"))?
+                    .clone();
+
+                entry
+                    .insert(KubeResource {
+                        resource: kube::api::ApiResource {
+                            group: resource.group.unwrap_or_default(),
+                            version: resource.version.unwrap_or_default(),
+                            api_version: api_version.to_string(),
+                            kind: kind.to_string(),
+                            plural: resource.name,
+                        },
+                        namespaced: resource.namespaced,
+                    })
+                    .clone()
+            }
+            hash_map::Entry::Occupied(entry) => entry.get().clone(),
```

Comment:
IIUC, we cache twice; once here in the client's `kube_resources`, and another time in the `pub list_resources_*()` (he `list_resources_*()` caches don't take into account the client).
Is that ok? Wouldn't it be ok with one of both?

Reply:
It's fine to not take the client into account when building this internal cache of `KubeResource` objects. We are always connected to the same Kubernetes cluster.

The only improvement I could make would be to put this HashMap behind a RwLock. Policy Server has multiple workers running policies, but all of them end up invoking the same host capabilities handler. There's the risk of having two concurrent inserts.

I'll fix that

Reply:
Done with 9d67a0b

Reply:
Good catch, thanks!